mpeg2/src/ts.rs

```diff
@@ -27,30 +30,112 @@ impl AdaptationField {
         if af_length as usize > buf.len() - 1 {
             return Err(DecodeError::new("adaptation field length too long"));
         } else if buf[0] > 0 {
+            let mut n = 2;
             af.discontinuity_indicator = Some(buf[1] & 0x80 != 0);
             af.random_access_indicator = Some(buf[1] & 0x40 != 0);
             af.program_clock_reference_27mhz = if af_length >= 7 && buf[1] & 0x10 != 0 {
                 let base = (buf[2] as u64) << 25 | (buf[3] as u64) << 17 | (buf[4] as u64) << 9 | (buf[5] as u64) << 1 | (buf[6] as u64) >> 7;
                 let ext = ((buf[6] as u64) & 1) << 8 | (buf[7] as u64);
+                n += 6;
                 Some(base * 300 + ext)
             } else {
                 None
             };
+
+            if buf[1] & 0b0000_1000 != 0 {
+                // skip OPCR
+                n += 6;
+            }
+            if buf[1] & 0b0000_0100 != 0 {
+                // skip splice_countdown
+                n += 1;
+            }
+            if buf[1] & 0b0000_0010 != 0 {
+                if n >= buf.len() || buf[n] as usize + n + 1 > buf.len() {
+                    return Err(DecodeError::new("adaptation field length too long"));
+                }
+                let transport_private_data_length = buf[n];
+                af.private_data_types = buf[n..(n + transport_private_data_length as usize)].to_vec();
+                n += transport_private_data_length as usize + 1
+            }
+            if buf[1] & 1 == 1 {
+                if n + 2 > buf.len() {
+                    return Err(DecodeError::new("adaptation field length too long"));
+                } else if buf[n] == 0 {
+                    return Err(DecodeError::new("invalid adaptation_field_extension_length"));
+                }
+                let temi_descriptors_len = buf[n] - 1;
```

Comment:
`buf[n]` is the `adaptation_field_extension_length` field, which includes the bytes required by the fields we skip below.

Let's change this variable so that it's clear that this is the length of the adaptation field extension and not just those descriptors:

```suggestion
                let adaptation_field_extension_length = buf[n];
```

Then fix the size of the buffer below.

Reply:
> I'm seeing a few more issues. Btw when the structures get as complex as this, I find it really helps to try to make the code look like the spec as much as possible. We've done a really good job of this in the h264 crate for example: https://github.com/sportsball-ai/av-rs/blob/main/h264/src/sequence_parameter_set.rs
> 
> It might be worth using some sort of bitstream reader like we do there so you can just write code like:
> 
> ```
> let adaptation_field_length = bitstream.read_u8()?;
> if (adaptation_field_length > 0) {
>     let discontinuity_indicator = bitstream.read_bit()?;
>     let random_access_indicator = bitstream.read_bit()?;
>     let elementary_stream_priority_indicator = bitstream.read_bit()?;
>     let pcr_flag = bitstream.read_bit()?;
>     let opcr_flag = bitstream.read_bit()?;
> 
>     ...
> 
>     if opcr_flag {
>         bitstream.skip_bytes(6)?;
>     }    
>     if splicing_point_flag {
>         bitstream.skip_bytes(1)?;
>     }    
> }
> ```
> 
> Then we can get rid of a large amount of the math and be able to verify correctness at a glance.

Hmm, this looks much nicer! I'll talk a look at h264 create, and try to make semi crate similar.